x-vector/src/X/Data/Vector/Generic.hs

```diff
@@ -165,6 +167,28 @@ uncons v
    in  (,) <$> (pre !? 0) <*> pure post
 {-# INLINE uncons #-}
 
+data IdxOff =
+  IdxOff !Int !Int
+
+unsafeSplits ::
+  Vector va a =>
+  Vector vb b =>
+  Vector vn Int =>
+  (va a -> b) ->
+  va a ->
+  vn Int ->
+  vb b
+unsafeSplits f xs sizes =
+  let
+    loop (IdxOff idx off) =
+      let
+        !len =
+          Generic.unsafeIndex sizes idx
+      in
+        Just (f $! Generic.unsafeSlice off len xs, IdxOff (idx + 1) (off + len))
+  in
+    Generic.unfoldrN (Generic.length sizes) loop (IdxOff 0 0)
+{-# INLINE unsafeSplits #-}
```

Comment:
yeah, something like
```
mapAccumulate
  (\off len ->
    (off + len, f $! Generic.unsafeSlice off len xs)
  )
  0
  sizes
```

hmm actually might not work since `mapAccumulate` requires input and output vectors to be same. never mind